mist-core/src/main/java/edu/snu/mist/core/master/lb/scaling/PeriodicDynamicScalingManager.java

```diff
@@ -0,0 +1,215 @@
+
+  /**
+   * The elapsed idle time.
+   */
+  private long idleTimeElapsed;
+
+  /**
+   * The elapsed overloaded time.
+   */
+  private long overloadedTimeElapsed;
+
+  /**
+   * The last measured timestamp.
+   */
+  private long lastMeasuredTimestamp;
+
+  /**
+   * Scheduled executor service for running periodic dynamic scaling.
+   */
+  private ScheduledExecutorService scheduledExecutorService;
+
+  @Inject
+  private PeriodicDynamicScalingManager(
+      final TaskStatsMap taskStatsMap,
+      @Parameter(DynamicScalingPeriod.class) final long dynamicScalingPeriod,
+      @Parameter(MaxTaskNum.class) final int maxTaskNum,
+      @Parameter(MinTaskNum.class) final int minTaskNum,
+      @Parameter(IdleTaskLoadThreshold.class) final double idleTaskLoadThreshold,
+      @Parameter(OverloadedTaskLoadThreshold.class) final double overloadedTaskLoadThreshold,
+      @Parameter(ScaleInGracePeriod.class) final long scaleInGracePeriod,
+      @Parameter(ScaleOutGracePeriod.class) final long scaleOutGracePeriod,
+      @Parameter(ScaleInIdleTaskRate.class) final double scaleInIdleTaskRate,
+      @Parameter(ScaleOutOverloadedTaskRate.class) final double scaleOutOverloadedTaskRate
+  ) {
+    this.taskStatsMap = taskStatsMap;
+    this.dynamicScalingPeriod = dynamicScalingPeriod;
+    this.maxTaskNum = maxTaskNum;
+    this.minTaskNum = minTaskNum;
+    this.idleTaskLoadThreshold = idleTaskLoadThreshold;
+    this.overloadedTaskLoadThreshold = overloadedTaskLoadThreshold;
+    this.scaleInGracePeriod = scaleInGracePeriod;
+    this.scaleOutGracePeriod = scaleOutGracePeriod;
+    this.scaleInIdleTaskRate = scaleInIdleTaskRate;
+    this.scaleOutOverloadedTaskRate = scaleOutOverloadedTaskRate;
+    this.idleTimeElapsed = 0L;
+    this.overloadedTimeElapsed = 0L;
+    this.lastMeasuredTimestamp = 0L;
+    this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  private boolean isClusterOverloaded() {
+    int overloadedTaskNum = 0;
+    for (final Map.Entry<String, TaskStats> entry : taskStatsMap.entrySet()) {
+      if (entry.getValue().getTaskLoad() > overloadedTaskLoadThreshold) {
+        overloadedTaskNum += 1;
+      }
+    }
+    final double overloadedTaskRate = overloadedTaskNum / taskStatsMap.getTaskList().size();
+    return overloadedTaskRate > scaleOutOverloadedTaskRate;
+  }
+
+  private boolean isClusterIdle() {
+    int idleTaskNum = 0;
+    for (final Map.Entry<String, TaskStats> entry : taskStatsMap.entrySet()) {
+      if (entry.getValue().getTaskLoad() <= idleTaskLoadThreshold) {
+        idleTaskNum += 1;
+      }
+    }
+    final double idleTaskRate = idleTaskNum / taskStatsMap.getTaskList().size();
+    return idleTaskRate > scaleInIdleTaskRate;
+  }
+
+  private final class AutoScaleRunner implements Runnable {
+
+    private AutoScaleRunner() {
+      // Do nothing.
+    }
+
+    @Override
+    public void run() {
+      final long oldTimeStamp;
+      if (lastMeasuredTimestamp == 0L) {
```

Comment:
We don't need this `if else` if we set the lastMeasuredTimestamp in the constructor